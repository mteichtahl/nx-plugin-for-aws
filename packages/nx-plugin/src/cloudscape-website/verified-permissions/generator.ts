/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  GeneratorCallback,
  OverwriteStrategy,
  Tree,
  generateFiles,
  installPackagesTask,
  joinPathFragments,
  readProjectConfiguration,
} from '@nx/devkit';
import { TsCloudscapeWebsiteVerifiedPermissionsGeneratorSchema } from './schema';
import {
  NxGeneratorInfo,
  getGeneratorInfo,
  readProjectConfigurationUnqualified,
} from '../../utils/nx';
import { addGeneratorMetricsIfApplicable } from '../../utils/metrics';
import { formatFilesInSubtree } from '../../utils/format';
import {
  PACKAGES_DIR,
  SHARED_CONSTRUCTS_DIR,
} from '../../utils/shared-constructs-constants';
import { addStarExport } from '../../utils/ast';

export const TS_CLOUDSCAPE_WEBSITE_VERIFIED_PERMISSIONS_GENERATOR_INFO: NxGeneratorInfo =
  getGeneratorInfo(__filename);

export const tsCloudscapeWebsiteVerifiedPermissionsGenerator = async (
  tree: Tree,
  options: TsCloudscapeWebsiteVerifiedPermissionsGeneratorSchema,
): Promise<GeneratorCallback> => {
  const principalEntityType = options.principalEntity;
  const namespace = options.namespace;

  // principalEntityType?
  // cedarJsonSchema? (location - we should then load this in if not provided create an empty one)
  // validateCedarSchema?

  const srcRoot = readProjectConfigurationUnqualified(
    tree,
    options.project,
  ).sourceRoot;

  const sourceTemplates = joinPathFragments(
    __dirname,
    'files',
    SHARED_CONSTRUCTS_DIR,
    'src',
    'core',
    'verified-permissions',
  );
  const destinationPath = joinPathFragments(
    PACKAGES_DIR,
    SHARED_CONSTRUCTS_DIR,
    'src',
    'core',
    'verified-permissions',
  );
  const templateValues = {
    namespace,
    principalEntityType,
  };

  // Check if verified permission has already been generated
  const isGenerated = !tree.exists(
    joinPathFragments(destinationPath, 'index.ts'),
  );

  if (!isGenerated) {
    throw new Error(
      `Verified permissions have already been generated for ${options.project}. Please delete the existing verified permissions before generating new ones.`,
    );
  }

  generateFiles(tree, sourceTemplates, destinationPath, templateValues, {
    overwriteStrategy: OverwriteStrategy.KeepExisting,
  });

  addStarExport(
    tree,
    joinPathFragments(destinationPath, '..', 'index.ts'),
    './verified-permissions/index.js',
  );

  await formatFilesInSubtree(tree);

  await addGeneratorMetricsIfApplicable(tree, [
    TS_CLOUDSCAPE_WEBSITE_VERIFIED_PERMISSIONS_GENERATOR_INFO,
  ]);

  await formatFilesInSubtree(tree);
  return () => {
    installPackagesTask(tree);
  };
};

export default tsCloudscapeWebsiteVerifiedPermissionsGenerator;
