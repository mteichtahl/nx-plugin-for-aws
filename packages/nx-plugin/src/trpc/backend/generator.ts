/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  addDependenciesToPackageJson,
  generateFiles,
  getPackageManagerCommand,
  installPackagesTask,
  joinPathFragments,
  OverwriteStrategy,
  ProjectConfiguration,
  Tree,
  updateJson,
} from '@nx/devkit';
import { TsTrpcApiGeneratorSchema } from './schema';
import { sharedConstructsGenerator } from '../../utils/shared-constructs';
import {
  PACKAGES_DIR,
  SHARED_CONSTRUCTS_DIR,
} from '../../utils/shared-constructs-constants';
import tsProjectGenerator from '../../ts/lib/generator';
import { getNpmScopePrefix, toScopeAlias } from '../../utils/npm-scope';
import { withVersions } from '../../utils/versions';
import { kebabCase, toClassName } from '../../utils/names';
import { formatFilesInSubtree } from '../../utils/format';
import { sortObjectKeys } from '../../utils/object';
import {
  NxGeneratorInfo,
  addGeneratorMetadata,
  getGeneratorInfo,
  readProjectConfigurationUnqualified,
} from '../../utils/nx';
import { addGeneratorMetricsIfApplicable } from '../../utils/metrics';
import { addApiGatewayConstruct } from '../../utils/api-constructs/api-constructs';
import { getLocalServerPortNumber } from '../../utils/port';
import { addZodV4Alias } from '../../utils/zod';

export const TRPC_BACKEND_GENERATOR_INFO: NxGeneratorInfo =
  getGeneratorInfo(__filename);

export async function tsTrpcApiGenerator(
  tree: Tree,
  options: TsTrpcApiGeneratorSchema,
) {
  await sharedConstructsGenerator(tree);

  const apiNamespace = getNpmScopePrefix(tree);
  const apiNameKebabCase = kebabCase(options.name);
  const apiNameClassName = toClassName(options.name);

  const backendName = apiNameKebabCase;
  const backendProjectName = `${apiNamespace}${backendName}`;

  const port = getLocalServerPortNumber(
    tree,
    TRPC_BACKEND_GENERATOR_INFO,
    2022,
  );

  await tsProjectGenerator(tree, {
    name: backendName,
    directory: options.directory,
  });

  const backendRoot = readProjectConfigurationUnqualified(
    tree,
    backendProjectName,
  ).root;

  const enhancedOptions = {
    backendProjectName,
    backendProjectAlias: toScopeAlias(backendProjectName),
    apiNameKebabCase,
    apiNameClassName,
    backendRoot,
    pkgMgrCmd: getPackageManagerCommand().exec,
    apiGatewayEventType: getApiGatewayEventType(options),
    port,
    ...options,
  };

  addApiGatewayConstruct(tree, {
    apiNameClassName,
    apiNameKebabCase,
    constructType:
      options.computeType === 'ServerlessApiGatewayHttpApi' ? 'http' : 'rest',
    backend: {
      type: 'trpc',
      projectAlias: enhancedOptions.backendProjectAlias,
      dir: backendRoot,
    },
    auth: options.auth,
  });

  updateJson(
    tree,
    joinPathFragments(backendRoot, 'project.json'),
    (config: ProjectConfiguration) => {
      config.metadata = {
        apiName: options.name,
        apiType: 'trpc',
        auth: options.auth,
        port,
      } as unknown;

      config.targets.serve = {
        executor: 'nx:run-commands',
        options: {
          commands: ['tsx --watch src/local-server.ts'],
          cwd: backendRoot,
        },
        continuous: true,
      };

      config.targets.bundle = {
        cache: true,
        executor: 'nx:run-commands',
        outputs: [`{workspaceRoot}/dist/${backendRoot}/bundle`],
        options: {
          command: `esbuild ${backendRoot}/src/router.ts --bundle --outfile=dist/${backendRoot}/bundle/index.js --platform=node --format=cjs`,
        },
      };
      config.targets.build.dependsOn = [
        ...(config.targets.build.dependsOn ?? []),
        'bundle',
      ];

      config.targets = sortObjectKeys(config.targets);
      return config;
    },
  );

  updateJson(
    tree,
    joinPathFragments(PACKAGES_DIR, SHARED_CONSTRUCTS_DIR, 'project.json'),
    (config: ProjectConfiguration) => {
      if (!config.targets) {
        config.targets = {};
      }
      if (!config.targets.build) {
        config.targets.build = {};
      }
      config.targets.build.dependsOn = [
        ...(config.targets.build.dependsOn ?? []),
        `${backendProjectName}:build`,
      ];
      return config;
    },
  );

  generateFiles(
    tree,
    joinPathFragments(__dirname, 'files'),
    backendRoot,
    enhancedOptions,
    {
      overwriteStrategy: OverwriteStrategy.Overwrite,
    },
  );

  tree.delete(joinPathFragments(backendRoot, 'src', 'lib'));

  // Add an alias for zod v4
  addZodV4Alias(tree);

  addDependenciesToPackageJson(
    tree,
    withVersions([
      'aws-xray-sdk-core',
      'zod',
      '@aws-lambda-powertools/logger',
      '@aws-lambda-powertools/metrics',
      '@aws-lambda-powertools/tracer',
      '@trpc/server',
      '@trpc/client',
      'aws4fetch',
      '@aws-sdk/credential-providers',
    ]),
    withVersions([
      '@types/aws-lambda',
      'esbuild',
      'tsx',
      'cors',
      '@types/cors',
    ]),
  );
  tree.delete(joinPathFragments(backendRoot, 'package.json'));

  addGeneratorMetadata(tree, backendName, TRPC_BACKEND_GENERATOR_INFO);

  await addGeneratorMetricsIfApplicable(tree, [TRPC_BACKEND_GENERATOR_INFO]);

  await formatFilesInSubtree(tree);
  return () => {
    installPackagesTask(tree);
  };
}

const getApiGatewayEventType = (options: TsTrpcApiGeneratorSchema): string => {
  if (options.computeType === 'ServerlessApiGatewayRestApi') {
    return 'APIGatewayProxyEvent';
  }
  if (options.auth === 'IAM') {
    return 'APIGatewayProxyEventV2WithIAMAuthorizer';
  } else if (options.auth === 'Cognito') {
    return 'APIGatewayProxyEventV2WithJWTAuthorizer';
  }
  return 'APIGatewayProxyEventV2';
};

export default tsTrpcApiGenerator;
