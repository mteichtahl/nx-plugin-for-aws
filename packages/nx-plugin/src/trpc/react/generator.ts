/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  addDependenciesToPackageJson,
  generateFiles,
  installPackagesTask,
  joinPathFragments,
  OverwriteStrategy,
  Tree,
} from '@nx/devkit';
import { ReactGeneratorSchema } from './schema';
import { JsxSelfClosingElement } from 'typescript';
import { runtimeConfigGenerator } from '../../cloudscape-website/runtime-config/generator';
import { toScopeAlias } from '../../utils/npm-scope';
import { withVersions } from '../../utils/versions';
import {
  createJsxElementFromIdentifier,
  replace,
  addSingleImport,
  query,
} from '../../utils/ast';
import { toClassName } from '../../utils/names';
import { formatFilesInSubtree } from '../../utils/format';
import {
  NxGeneratorInfo,
  getGeneratorInfo,
  readProjectConfigurationUnqualified,
} from '../../utils/nx';
import { addGeneratorMetricsIfApplicable } from '../../utils/metrics';
import { addTargetToServeLocal } from '../../api-connection/serve-local';

export const TRPC_REACT_GENERATOR_INFO: NxGeneratorInfo =
  getGeneratorInfo(__filename);

export async function reactGenerator(
  tree: Tree,
  options: ReactGeneratorSchema,
) {
  const frontendProjectConfig = readProjectConfigurationUnqualified(
    tree,
    options.frontendProjectName,
  );
  const backendProjectConfig = readProjectConfigurationUnqualified(
    tree,
    options.backendProjectName,
  );
  /* eslint-disable @typescript-eslint/no-explicit-any */
  const metadata = backendProjectConfig.metadata as any;
  const apiName = metadata.apiName;
  const auth = metadata.auth ?? 'IAM';
  const port = metadata.port ?? 2022;
  const apiNameClassName = toClassName(apiName);
  const backendProjectAlias = toScopeAlias(backendProjectConfig.name);

  generateFiles(
    tree,
    joinPathFragments(__dirname, 'files'),
    frontendProjectConfig.root,
    {
      apiName,
      apiNameClassName: toClassName(apiName),
      ...options,
      auth,
      backendProjectAlias,
    },
    {
      overwriteStrategy: OverwriteStrategy.KeepExisting,
    },
  );

  // Generate the tanstack query provider if it does not exist already
  if (
    !tree.exists(
      joinPathFragments(
        frontendProjectConfig.sourceRoot,
        'components',
        'QueryClientProvider.tsx',
      ),
    )
  ) {
    generateFiles(
      tree,
      joinPathFragments(
        __dirname,
        '../../utils/files/website/components/tanstack-query',
      ),
      joinPathFragments(frontendProjectConfig.sourceRoot, 'components'),
      {},
    );
  }

  if (auth === 'IAM') {
    generateFiles(
      tree,
      joinPathFragments(__dirname, '../../utils/files/website/hooks/sigv4'),
      joinPathFragments(frontendProjectConfig.sourceRoot, 'hooks'),
      {},
    );
  }

  await runtimeConfigGenerator(tree, {
    project: frontendProjectConfig.name,
  });

  // update main.tsx
  const mainTsxPath = joinPathFragments(
    frontendProjectConfig.sourceRoot,
    'main.tsx',
  );
  addSingleImport(
    tree,
    mainTsxPath,
    'QueryClientProvider',
    './components/QueryClientProvider',
  );

  const clientProviderName = `${apiNameClassName}ClientProvider`;
  addSingleImport(
    tree,
    mainTsxPath,
    clientProviderName,
    `./components/${clientProviderName}`,
  );

  // Check if QueryClientProvider already exists
  const hasQueryClientProvider =
    query(
      tree,
      mainTsxPath,
      'JsxOpeningElement[tagName.name="QueryClientProvider"]',
    ).length > 0;

  if (!hasQueryClientProvider) {
    replace(
      tree,
      mainTsxPath,
      'JsxSelfClosingElement[tagName.name="App"]',
      (node: JsxSelfClosingElement) =>
        createJsxElementFromIdentifier('QueryClientProvider', [node]),
    );
  }

  // Check if client provider already exists
  const hasProvider =
    query(
      tree,
      mainTsxPath,
      `JsxOpeningElement[tagName.name="${clientProviderName}"]`,
    ).length > 0;
  if (!hasProvider) {
    replace(
      tree,
      mainTsxPath,
      'JsxSelfClosingElement[tagName.name="App"]',
      (node: JsxSelfClosingElement) =>
        createJsxElementFromIdentifier(clientProviderName, [node]),
    );
  }

  addTargetToServeLocal(
    tree,
    frontendProjectConfig.name,
    backendProjectConfig.name,
    {
      url: `http://localhost:${port}/`,
      apiName,
    },
  );

  addDependenciesToPackageJson(
    tree,
    withVersions([
      '@trpc/client',
      '@trpc/tanstack-react-query',
      '@tanstack/react-query',
      ...((auth === 'IAM'
        ? [
            'oidc-client-ts',
            'aws4fetch',
            '@aws-sdk/client-cognito-identity',
            '@aws-sdk/credential-provider-cognito-identity',
            'react-oidc-context',
          ]
        : []) as any),
      ...((auth === 'Cognito' ? ['react-oidc-context'] : []) as any),
    ]),
    withVersions(['@smithy/types']),
  );

  await addGeneratorMetricsIfApplicable(tree, [TRPC_REACT_GENERATOR_INFO]);

  await formatFilesInSubtree(tree);
  return () => {
    installPackagesTask(tree);
  };
}
export default reactGenerator;
