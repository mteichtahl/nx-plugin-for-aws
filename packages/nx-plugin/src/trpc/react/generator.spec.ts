/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { Tree, updateJson } from '@nx/devkit';
import { TRPC_REACT_GENERATOR_INFO, reactGenerator } from './generator';
import { createTreeUsingTsSolutionSetup } from '../../utils/test';
import { sharedConstructsGenerator } from '../../utils/shared-constructs';
import { expectHasMetricTags } from '../../utils/metrics.spec';
import { tsCloudScapeWebsiteGenerator } from '../../cloudscape-website/app/generator';
import { tsTrpcApiGenerator } from '../backend/generator';

describe('trpc react generator', () => {
  let tree: Tree;

  beforeEach(() => {
    tree = createTreeUsingTsSolutionSetup();
    // Mock frontend project configuration
    tree.write(
      'apps/frontend/project.json',
      JSON.stringify({
        name: 'frontend',
        root: 'apps/frontend',
        sourceRoot: 'apps/frontend/src',
      }),
    );
    // Mock backend project configuration
    tree.write(
      'apps/backend/project.json',
      JSON.stringify({
        name: 'backend',
        root: 'apps/backend',
        sourceRoot: 'apps/backend/src',
        metadata: {
          apiName: 'TestApi',
          auth: 'None',
        },
      }),
    );
    // Mock main.tsx file
    tree.write(
      'apps/frontend/src/main.tsx',
      `
import { RouterProvider } from '@tanstack/react-router';

const App = () => <RouterProvider router={router} />;

export function Main() {
  return <App />;
}
`,
    );
  });

  it('should generate trpc react files', async () => {
    await reactGenerator(tree, {
      frontendProjectName: 'frontend',
      backendProjectName: 'backend',
    });
    // Verify generated files
    expect(
      tree.exists('apps/frontend/src/components/TestApiClientProvider.tsx'),
    ).toBeTruthy();

    expect(
      tree.read(
        'apps/frontend/src/components/TestApiClientProvider.tsx',
        'utf-8',
      ),
    ).toMatchSnapshot('TestApiClientProvider.tsx');

    expect(tree.exists('apps/frontend/src/hooks/useTestApi.tsx')).toBeTruthy();
    expect(
      tree.read('apps/frontend/src/hooks/useTestApi.tsx', 'utf-8'),
    ).toMatchSnapshot('useTestApi.tsx');
  });

  it('should modify main.tsx correctly', async () => {
    await reactGenerator(tree, {
      frontendProjectName: 'frontend',
      backendProjectName: 'backend',
    });
    const mainTsxContent = tree.read('apps/frontend/src/main.tsx', 'utf-8');
    // Create snapshot of modified main.tsx
    expect(mainTsxContent).toMatchSnapshot('main.tsx');
  });

  it('should add required dependencies', async () => {
    await reactGenerator(tree, {
      frontendProjectName: 'frontend',
      backendProjectName: 'backend',
    });
    const packageJson = JSON.parse(tree.read('package.json', 'utf-8'));
    // Verify dependencies were added
    expect(
      packageJson.dependencies['@trpc/tanstack-react-query'],
    ).toBeDefined();
    expect(packageJson.dependencies['@tanstack/react-query']).toBeDefined();
  });

  it('should handle IAM auth option', async () => {
    updateJson(tree, 'apps/backend/project.json', (config) => ({
      ...config,
      metadata: {
        ...config.metadata,
        auth: 'IAM',
      },
    }));

    await reactGenerator(tree, {
      frontendProjectName: 'frontend',
      backendProjectName: 'backend',
    });

    expect(
      tree.read(
        'apps/frontend/src/components/TestApiClientProvider.tsx',
        'utf-8',
      ),
    ).toMatchSnapshot('TestApiClientProvider-IAM.tsx');

    expect(tree.exists('apps/frontend/src/hooks/useSigV4.tsx')).toBeTruthy();
    expect(
      tree.read('apps/frontend/src/hooks/useSigV4.tsx', 'utf-8'),
    ).toMatchSnapshot('useSigV4.tsx');

    const packageJson = JSON.parse(tree.read('package.json', 'utf-8'));
    // Verify dependencies were added
    expect(
      packageJson.dependencies['@trpc/tanstack-react-query'],
    ).toBeDefined();
    expect(packageJson.dependencies['@tanstack/react-query']).toBeDefined();
    expect(packageJson.dependencies['oidc-client-ts']).toBeDefined();
    expect(packageJson.dependencies['react-oidc-context']).toBeDefined();
    expect(
      packageJson.dependencies['@aws-sdk/client-cognito-identity'],
    ).toBeDefined();
    expect(
      packageJson.dependencies['@aws-sdk/credential-provider-cognito-identity'],
    ).toBeDefined();
    expect(packageJson.dependencies['aws4fetch']).toBeDefined();
  });

  it('should handle Cognito auth option', async () => {
    updateJson(tree, 'apps/backend/project.json', (config) => ({
      ...config,
      metadata: {
        ...config.metadata,
        auth: 'Cognito',
      },
    }));

    await reactGenerator(tree, {
      frontendProjectName: 'frontend',
      backendProjectName: 'backend',
    });

    expect(
      tree.read(
        'apps/frontend/src/components/TestApiClientProvider.tsx',
        'utf-8',
      ),
    ).toMatchSnapshot('TestApiClientProvider-Cognito.tsx');

    expect(tree.exists('apps/frontend/src/hooks/useSigV4.tsx')).toBeFalsy();

    const packageJson = JSON.parse(tree.read('package.json', 'utf-8'));
    // Verify dependencies were added
    expect(
      packageJson.dependencies['@trpc/tanstack-react-query'],
    ).toBeDefined();
    expect(packageJson.dependencies['@tanstack/react-query']).toBeDefined();
    expect(packageJson.dependencies['react-oidc-context']).toBeDefined();
  });

  it('should add generator metric to app.ts', async () => {
    await sharedConstructsGenerator(tree);

    // Call the generator function
    await reactGenerator(tree, {
      frontendProjectName: 'frontend',
      backendProjectName: 'backend',
    });

    // Verify the metric was added to app.ts
    expectHasMetricTags(tree, TRPC_REACT_GENERATOR_INFO.metric);
  });
});

describe('trpc react generator with unqualified names', () => {
  let tree: Tree;

  beforeEach(() => {
    tree = createTreeUsingTsSolutionSetup();

    // Setup package.json with a scope
    tree.write(
      'package.json',
      JSON.stringify({
        name: '@my-scope/monorepo',
        version: '1.0.0',
      }),
    );

    // Mock frontend project configuration with TypeScript fully qualified name
    tree.write(
      'apps/frontend/project.json',
      JSON.stringify({
        name: '@my-scope/frontend',
        root: 'apps/frontend',
        sourceRoot: 'apps/frontend/src',
      }),
    );

    // Mock backend project configuration with TypeScript fully qualified name
    tree.write(
      'apps/backend/project.json',
      JSON.stringify({
        name: '@my-scope/backend',
        root: 'apps/backend',
        sourceRoot: 'apps/backend/src',
        metadata: {
          apiName: 'TestApi',
          auth: 'None',
        },
      }),
    );

    // Mock main.tsx file
    tree.write(
      'apps/frontend/src/main.tsx',
      `
import { RouterProvider } from '@tanstack/react-router';

const App = () => <RouterProvider router={router} />;

export function Main() {
  return <App />;
}
`,
    );
  });

  it('should work with unqualified frontend project name', async () => {
    await reactGenerator(tree, {
      frontendProjectName: 'frontend', // Unqualified name (without @my-scope/)
      backendProjectName: '@my-scope/backend', // Fully qualified name
    });

    // Verify files were generated
    expect(
      tree.exists('apps/frontend/src/components/TestApiClientProvider.tsx'),
    ).toBeTruthy();
    expect(tree.exists('apps/frontend/src/hooks/useTestApi.tsx')).toBeTruthy();
  });

  it('should work with unqualified backend project name', async () => {
    await reactGenerator(tree, {
      frontendProjectName: '@my-scope/frontend', // Fully qualified name
      backendProjectName: 'backend', // Unqualified name (without @my-scope/)
    });

    // Verify files were generated
    expect(
      tree.exists('apps/frontend/src/components/TestApiClientProvider.tsx'),
    ).toBeTruthy();
    expect(tree.exists('apps/frontend/src/hooks/useTestApi.tsx')).toBeTruthy();
  });

  it('should work with both unqualified project names', async () => {
    await reactGenerator(tree, {
      frontendProjectName: 'frontend', // Unqualified name (without @my-scope/)
      backendProjectName: 'backend', // Unqualified name (without @my-scope/)
    });

    // Verify files were generated
    expect(
      tree.exists('apps/frontend/src/components/TestApiClientProvider.tsx'),
    ).toBeTruthy();
    expect(tree.exists('apps/frontend/src/hooks/useTestApi.tsx')).toBeTruthy();
  });
});

describe('trpc react generator with real react and trpc projects', () => {
  let tree: Tree;

  beforeEach(async () => {
    tree = createTreeUsingTsSolutionSetup();

    // Generate a cloudscape website
    await tsCloudScapeWebsiteGenerator(tree, {
      name: 'frontend',
      skipInstall: true,
    });
  });

  it('should configure serve-local integration with generated projects', async () => {
    // Generate a trpc backend
    await tsTrpcApiGenerator(tree, {
      name: 'TestApi',
      auth: 'None',
      computeType: 'ServerlessApiGatewayHttpApi',
    });

    await reactGenerator(tree, {
      frontendProjectName: 'frontend',
      backendProjectName: 'test-api',
    });

    // Read the frontend project configuration
    const frontendProject = JSON.parse(
      tree.read('frontend/project.json', 'utf-8'),
    );

    // Verify that serve-local target now depends on backend serve target
    expect(frontendProject.targets['serve-local'].dependsOn).toContainEqual({
      projects: ['@proj/test-api'],
      target: 'serve',
    });

    // Verify that the runtime config was created and modified
    expect(
      tree.exists('frontend/src/components/RuntimeConfig/index.tsx'),
    ).toBeTruthy();

    const runtimeConfigContent = tree.read(
      'frontend/src/components/RuntimeConfig/index.tsx',
      'utf-8',
    );

    // Verify that the runtime config includes the API override
    expect(runtimeConfigContent).toContain('runtimeConfig.apis.TestApi');
    expect(runtimeConfigContent).toContain('http://localhost:2022/');
  });

  it('should use correct port numbers in runtime config overrides', async () => {
    // Generate first API
    await tsTrpcApiGenerator(tree, {
      name: 'FirstApi',
      auth: 'None',
      computeType: 'ServerlessApiGatewayHttpApi',
    });

    // Generate second API
    await tsTrpcApiGenerator(tree, {
      name: 'SecondApi',
      auth: 'None',
      computeType: 'ServerlessApiGatewayHttpApi',
    });

    // Connect first API to frontend
    await reactGenerator(tree, {
      frontendProjectName: 'frontend',
      backendProjectName: 'first-api',
    });

    // Connect second API to frontend
    await reactGenerator(tree, {
      frontendProjectName: 'frontend',
      backendProjectName: 'second-api',
    });

    // Verify that the runtime config includes the correct port overrides
    const runtimeConfigContent = tree.read(
      'frontend/src/components/RuntimeConfig/index.tsx',
      'utf-8',
    );

    expect(runtimeConfigContent).toContain('runtimeConfig.apis.FirstApi');
    expect(runtimeConfigContent).toContain('http://localhost:2022/');
    expect(runtimeConfigContent).toContain('runtimeConfig.apis.SecondApi');
    expect(runtimeConfigContent).toContain('http://localhost:2023/');
  });
});
