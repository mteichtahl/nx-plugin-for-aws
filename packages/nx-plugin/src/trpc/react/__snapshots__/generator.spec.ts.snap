// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`trpc react generator > should generate trpc react files > TestApiClientProvider.tsx 1`] = `
"import { AppRouter } from 'backend';
import { useQueryClient } from '@tanstack/react-query';
import { createTRPCContext } from '@trpc/tanstack-react-query';
import { FC, PropsWithChildren, useMemo } from 'react';
import { useRuntimeConfig } from '../hooks/useRuntimeConfig';
import { HTTPLinkOptions, createTRPCClient, httpLink } from '@trpc/client';

const { TRPCProvider, useTRPC } = createTRPCContext<AppRouter>();

export const useTestApi = useTRPC;

export const TestApiClientProvider: FC<PropsWithChildren> = ({ children }) => {
  const queryClient = useQueryClient();
  const runtimeConfig = useRuntimeConfig();
  const apiUrl = runtimeConfig.apis.TestApi;

  const trpcClient = useMemo(() => {
    const linkOptions: HTTPLinkOptions<any> = {
      url: apiUrl,
    };

    return createTRPCClient<AppRouter>({
      links: [httpLink(linkOptions)],
    });
  }, [apiUrl]);

  return (
    <TRPCProvider trpcClient={trpcClient} queryClient={queryClient}>
      {children}
    </TRPCProvider>
  );
};

export default TestApiClientProvider;
"
`;

exports[`trpc react generator > should generate trpc react files > useTestApi.tsx 1`] = `
"import { useTestApi as useClient } from '../components/TestApiClientProvider';

export const useTestApi = useClient;
"
`;

exports[`trpc react generator > should handle Cognito auth option > TestApiClientProvider-Cognito.tsx 1`] = `
"import { AppRouter } from 'backend';
import { useQueryClient } from '@tanstack/react-query';
import { createTRPCContext } from '@trpc/tanstack-react-query';
import { FC, PropsWithChildren, useMemo } from 'react';
import { useRuntimeConfig } from '../hooks/useRuntimeConfig';
import { HTTPLinkOptions, createTRPCClient, httpLink } from '@trpc/client';
import { useAuth } from 'react-oidc-context';

const { TRPCProvider, useTRPC } = createTRPCContext<AppRouter>();

export const useTestApi = useTRPC;

export const TestApiClientProvider: FC<PropsWithChildren> = ({ children }) => {
  const queryClient = useQueryClient();
  const runtimeConfig = useRuntimeConfig();
  const apiUrl = runtimeConfig.apis.TestApi;

  const { user } = useAuth();

  const trpcClient = useMemo(() => {
    const linkOptions: HTTPLinkOptions<any> = {
      url: apiUrl,
      headers: {
        Authorization: \`Bearer \${user?.id_token}\`,
      },
    };

    return createTRPCClient<AppRouter>({
      links: [httpLink(linkOptions)],
    });
  }, [apiUrl, user]);

  return (
    <TRPCProvider trpcClient={trpcClient} queryClient={queryClient}>
      {children}
    </TRPCProvider>
  );
};

export default TestApiClientProvider;
"
`;

exports[`trpc react generator > should handle IAM auth option > TestApiClientProvider-IAM.tsx 1`] = `
"import { AppRouter } from 'backend';
import { useQueryClient } from '@tanstack/react-query';
import { createTRPCContext } from '@trpc/tanstack-react-query';
import { FC, PropsWithChildren, useMemo } from 'react';
import { useRuntimeConfig } from '../hooks/useRuntimeConfig';
import { HTTPLinkOptions, createTRPCClient, httpLink } from '@trpc/client';
import { useSigV4 } from '../hooks/useSigV4';

const { TRPCProvider, useTRPC } = createTRPCContext<AppRouter>();

export const useTestApi = useTRPC;

export const TestApiClientProvider: FC<PropsWithChildren> = ({ children }) => {
  const queryClient = useQueryClient();
  const runtimeConfig = useRuntimeConfig();
  const apiUrl = runtimeConfig.apis.TestApi;

  const sigv4Client = useSigV4();

  const trpcClient = useMemo(() => {
    const linkOptions: HTTPLinkOptions<any> = {
      url: apiUrl,
      fetch: sigv4Client,
    };

    return createTRPCClient<AppRouter>({
      links: [httpLink(linkOptions)],
    });
  }, [apiUrl, sigv4Client]);

  return (
    <TRPCProvider trpcClient={trpcClient} queryClient={queryClient}>
      {children}
    </TRPCProvider>
  );
};

export default TestApiClientProvider;
"
`;

exports[`trpc react generator > should handle IAM auth option > useSigV4.tsx 1`] = `
"import { AwsClient } from 'aws4fetch';
import { CognitoIdentityClient } from '@aws-sdk/client-cognito-identity';
import { fromCognitoIdentityPool } from '@aws-sdk/credential-provider-cognito-identity';
import { useCallback, useState } from 'react';
import { useAuth } from 'react-oidc-context';
import { useRuntimeConfig } from './useRuntimeConfig';
import {
  AwsCredentialIdentity,
  AwsCredentialIdentityProvider,
} from '@smithy/types';

// Credential expiration grace time before considering credentials as expired
const CREDENTIAL_EXPIRY_OFFSET_MILLIS = 30 * 1000;

export const useSigV4 = () => {
  const { cognitoProps } = useRuntimeConfig();
  const { user } = useAuth();

  const [cachedCredentials, setCachedCredentials] = useState<{
    [key: string]: AwsCredentialIdentity;
  }>({});

  const withCachedCredentials = useCallback(
    async (
      provider: AwsCredentialIdentityProvider,
      ...cacheKeys: string[]
    ): Promise<AwsCredentialIdentity> => {
      const key = \`sigv4/\${cacheKeys.join('/')}\`;
      const cachedCreds = cachedCredentials[key];
      if (
        cachedCreds &&
        cachedCreds.expiration &&
        cachedCreds.expiration.getTime() >
          Date.now() + CREDENTIAL_EXPIRY_OFFSET_MILLIS
      ) {
        return cachedCreds;
      }
      const credentials = await provider();
      setCachedCredentials((prev) => ({ ...prev, [key]: credentials }));
      return credentials;
    },
    [cachedCredentials, setCachedCredentials],
  );

  return useCallback(
    async (input: RequestInfo | URL, init?: RequestInit | undefined) => {
      if (!cognitoProps && import.meta.env.MODE === 'serve-local') {
        // Skip request signing in serve-local mode when cognitoProps are not set
        return fetch(input, init);
      }
      if (!cognitoProps) {
        throw new Error('cognitoProps is undefined!');
      }
      if (!user?.id_token) {
        throw new Error('user.id_token is undefined!');
      }

      const credentialsFromCognitoIdentityPool = fromCognitoIdentityPool({
        client: new CognitoIdentityClient({ region: cognitoProps.region }),
        identityPoolId: cognitoProps.identityPoolId,
        logins: {
          [\`cognito-idp.\${cognitoProps.region}.amazonaws.com/\${cognitoProps.userPoolId}\`]:
            user.id_token,
        },
      });
      const cognitoidentity = new CognitoIdentityClient({
        credentials: credentialsFromCognitoIdentityPool,
      });
      const credential = await withCachedCredentials(
        cognitoidentity.config.credentials,
        cognitoProps.identityPoolId,
        user.profile.sub,
      );
      const awsClient = new AwsClient(credential);
      return awsClient.fetch(input, init);
    },
    [cognitoProps, user?.id_token, user?.profile.sub, withCachedCredentials],
  );
};
"
`;

exports[`trpc react generator > should modify main.tsx correctly > main.tsx 1`] = `
"import TestApiClientProvider from './components/TestApiClientProvider';
import QueryClientProvider from './components/QueryClientProvider';
import RuntimeConfigProvider from './components/RuntimeConfig';
import { RouterProvider } from '@tanstack/react-router';

const App = () => <RouterProvider router={router} />;

export function Main() {
  return (
    <RuntimeConfigProvider>
      <QueryClientProvider>
        <TestApiClientProvider>
          <App />
        </TestApiClientProvider>
      </QueryClientProvider>
    </RuntimeConfigProvider>
  );
}
"
`;
